File: scripts/split-code-diagrams.py
#!/usr/bin/env python3
"""
Split Code and Diagram Slides
-------------------------------
Separates code blocks and Mermaid diagrams from bullet points
to prevent slide overflow.

Usage:
    python3 scripts/split-code-diagrams.py <path/to/slides-data.json>
    python3 scripts/split-code-diagrams.py --all  # Process all presentations

For each slide with both content and code/diagram:
- Slide A: Title + bullet points (explanation)
- Slide B: Title + "(コード例)" or "(図解)" + code/diagram only
"""

import json
import sys
from pathlib import Path


def split_slide_content(slides):
    """Split slides with both content and code/diagram into separate slides."""
    new_slides = []
    split_count = 0

    for slide in slides:
        has_mermaid = 'mermaid' in slide and slide.get('mermaid')
        has_code = 'code' in slide and slide.get('code')
        has_content = 'content' in slide and len(slide.get('content', [])) > 0

        # If slide has code/diagram AND content, split it
        if (has_mermaid or has_code) and has_content:
            split_count += 1

            # Slide 1: Explanation only
            slide1 = {
                "title": slide['title'],
                "content": slide['content']
            }
            new_slides.append(slide1)

            # Slide 2: Code/diagram only
            if has_mermaid:
                # Mermaid diagram
                slide2 = {
                    "title": slide['title'] + "（図解）",
                    "content": [],
                    "mermaid": slide['mermaid']
                }
            elif has_code:
                # Code block
                slide2 = {
                    "title": slide['title'] + "（コード例）",
                    "content": [],
                    "code": slide['code'],
                    "codeLanguage": slide.get('codeLanguage', 'text')
                }

            new_slides.append(slide2)
        else:
            # No split needed
            new_slides.append(slide)

    return new_slides, split_count


def process_file(json_path):
    """Process a single slides-data.json file."""
    json_path = Path(json_path)

    if not json_path.exists():
        print(f"❌ File not found: {json_path}")
        return False

    print(f"📄 Processing: {json_path}")

    # Load JSON
    with open(json_path, 'r', encoding='utf-8') as f:
        data = json.load(f)

    original_count = len(data['slides'])

    # Split slides
    new_slides, split_count = split_slide_content(data['slides'])
    data['slides'] = new_slides

    # Save updated JSON
    with open(json_path, 'w', encoding='utf-8') as f:
        json.dump(data, f, ensure_ascii=False, indent=2)

    print(f"  ✅ Split {split_count} slides")
    print(f"  📊 {original_count} → {len(new_slides)} slides\n")

    return True


def process_all_presentations():
    """Process all presentations in docs/ directory."""
    docs_dir = Path(__file__).parent.parent / 'docs'

    if not docs_dir.exists():
        print(f"❌ docs/ directory not found")
        return

    print("🔍 Searching for presentations in docs/...\n")

    total = 0
    success = 0

    for presentation_dir in sorted(docs_dir.iterdir()):
        if not presentation_dir.is_dir():
            continue

        json_path = presentation_dir / 'slides-data.json'

        if not json_path.exists():
            continue

        total += 1
        if process_file(json_path):
            success += 1

    print(f"📊 Summary: {success}/{total} presentations processed")


def main():
    if len(sys.argv) < 2:
        print("Usage:")
        print("  python3 scripts/split-code-diagrams.py <path/to/slides-data.json>")
        print("  python3 scripts/split-code-diagrams.py --all")
        sys.exit(1)

    if sys.argv[1] == '--all':
        process_all_presentations()
    else:
        json_path = sys.argv[1]
        if process_file(json_path):
            sys.exit(0)
        else:
            sys.exit(1)


if __name__ == '__main__':
    main()

File: scripts/split-code-diagrams.ts
#!/usr/bin/env bun
/**
 * Split slides that have both bullet content AND a code block into two slides.
 * Replaces scripts/split-code-diagrams.py
 *
 * Slide A: Title + bullet points (explanation) — preserves subtitle, speakerNotes, layout
 * Slide B: Title + "（コード例）" + code block only
 *
 * Usage:
 *   bun run scripts/split-code-diagrams.ts --all
 *   bun run scripts/split-code-diagrams.ts docs/my-deck/slides-data.json
 */

import { Glob } from "bun";
// MAX_BULLET_CHARS is available from "./lib/constants.js" for future use in overflow checks.

interface Slide {
	title?: string;
	subtitle?: string;
	content?: string[];
	code?: string;
	codeLanguage?: string;
	speakerNotes?: string;
	layout?: string;
}

interface SlideData {
	slides: Slide[];
}

function splitCodeDiagrams(slides: Slide[]): {
	slides: Slide[];
	splitCount: number;
} {
	const newSlides: Slide[] = [];
	let splitCount = 0;

	for (const slide of slides) {
		const hasCode = !!slide.code;
		const hasContent = (slide.content ?? []).length > 0;

		if (hasCode && hasContent) {
			splitCount++;

			// Slide 1: explanation only — preserve subtitle, speakerNotes, layout
			const { code: _code, codeLanguage: _lang, ...slideBase } = slide;
			const slide1: Slide = { ...slideBase };
			newSlides.push(slide1);

			// Slide 2: code block only — no subtitle/speakerNotes
			const slide2: Slide = {
				title: `${slide.title ?? ""}（コード例）`,
				content: [],
				code: slide.code,
				codeLanguage: slide.codeLanguage ?? "text",
				layout: slide.layout,
			};
			newSlides.push(slide2);
		} else {
			newSlides.push(slide);
		}
	}

	return { slides: newSlides, splitCount };
}

async function processFile(filePath: string): Promise<boolean> {
	const file = Bun.file(filePath);
	if (!(await file.exists())) {
		console.error(`❌ File not found: ${filePath}`);
		return false;
	}

	console.log(`📄 Processing: ${filePath}`);

	try {
		const data = (await file.json()) as SlideData;
		const originalCount = data.slides.length;
		const { slides, splitCount } = splitCodeDiagrams(data.slides);
		data.slides = slides;

		await Bun.write(filePath, JSON.stringify(data, null, 2));
		console.log(`  ✅ Split ${splitCount} slides`);
		console.log(`  📊 ${originalCount} → ${slides.length} slides\n`);

		// Update slides.count in config to reflect new slide count
		const configPath = filePath.replace(
			"slides-data.json",
			"slides.config.yaml",
		);
		if (await Bun.file(configPath).exists()) {
			try {
				const { parse: parseYaml, stringify: stringifyYaml } = await import(
					"yaml"
				);
				const configText = await Bun.file(configPath).text();
				const config = parseYaml(configText) as Record<string, unknown>;
				if (config.slides && typeof config.slides === "object") {
					(config.slides as Record<string, unknown>).count = slides.length;
				} else {
					config.slides = { count: slides.length };
				}
				await Bun.write(configPath, stringifyYaml(config));
			} catch {
				// Config update is best-effort — don't fail the split
			}
		}

		return true;
	} catch (err) {
		console.error(
			`❌ Error processing ${filePath}: ${err instanceof Error ? err.message : String(err)}`,
		);
		return false;
	}
}

async function main() {
	const args = process.argv.slice(2);

	if (args[0] === "--all") {
		const glob = new Glob("docs/**/slides-data.json");
		const files = Array.from(glob.scanSync()).sort();
		let success = 0;
		const total = files.length;
		for (const f of files) {
			if (await processFile(f)) success++;
		}
		console.log(`📊 Summary: ${success}/${total} presentations processed`);
	} else if (args[0]) {
		const ok = await processFile(args[0]);
		process.exit(ok ? 0 : 1);
	} else {
		console.error("Usage:");
		console.error("  bun run scripts/split-code-diagrams.ts --all");
		console.error(
			"  bun run scripts/split-code-diagrams.ts docs/my-deck/slides-data.json",
		);
		process.exit(1);
	}
}

main();
